### ISSUE TITLE
Add `[DoesNotReturnWhenNull(nameof(parameter))]`

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#37544]
Currently, we have `DoesNotReturn` and `DoesNotReturnIf(bool)`. The request is for `DoesNotReturnWhenNull(string)` or something to that effect.

Relates to https://github.com/dotnet/corefx/issues/37826, https://github.com/dotnet/roslyn/issues/35816#issuecomment-504151108, and https://github.com/dotnet/csharplang/issues/2536#issuecomment-493608421 (request for this attribute was raised in the threads)

Relates to the design of annotation attributes (see [LDM notes](https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-05-15.md))

**Comment by RikkiGibson** [comment]
I seem to remember we wanted to suggest something like the following for this scenario:

```cs
public void EnsureNotNull<T>([NotNull] T arg) where T : class?;
```

This way, we would accept a maybe-null argument and understand after the method returns that the argument is non-null.

**Comment by jcouv** [comment]
Good point. This should already work.
I'll go ahead and close this

**Comment by jcouv** [comment]
@RikkiGibson It turns out that `[NotNull]` does not work for this purpose. Below is a test demonstrating the issue. Re-opening

```C#
        [Fact]
        public void Test()
        {
            var source = @"
using System;
#nullable enable

public class C
{
    void M2(string? x)
    {
        M(x);
        x.ToString(); // warning
    }
    void M([NotNull] string? x) // when this method returns, `x` was not null
    {
        if (x is null) throw null!;
    }
}

[AttributeUsage(AttributeTargets.Field | AttributeTargets.Parameter | AttributeTargets.Property | AttributeTargets.ReturnValue)]
public sealed class NotNullAttribute : Attribute
{
}";
            var comp = CreateNullableCompilation(source);
            comp.VerifyDiagnostics(
                    // (10,9): warning CS8602: Dereference of a possibly null reference.
                    //         x.ToString(); // warning
                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(10, 9)
                    );
        }
```

**Comment by jcouv** [comment]
Hum, I'm confused. I do see code that should learn from post-condition attributes on input parameters, but the observed behavior doesn't match that...

```C#
            switch (refKind)
            {
                case RefKind.None:
                case RefKind.In:
                    {
                        // learn from post-conditions [Maybe/NotNull, Maybe/NotNullWhen] without using an assignment
                        learnFromPostConditions(argument, parameterAnnotations);
                    }
                    break;
```

**Comment by jcouv** [comment]
Turns out the code does learn from post-conditions on inputs. My test above was incorrect because the attribute was not the proper namespace. Here's the current behavior:

```C#

        [Fact]
        public void Test()
        {
            var source = @"
using System;
using System.Diagnostics.CodeAnalysis;
#nullable enable

public class C
{
    void M2(string? x)
    {
        M(x);
        x.ToString(); // no warning
    }
    void M([NotNull] string? x) // when this method returns, `x` was not null
    {
        if (x is null) throw null!;
    }
}

namespace System.Diagnostics.CodeAnalysis
{
    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Parameter | AttributeTargets.Property | AttributeTargets.ReturnValue)]
    public sealed class NotNullAttribute : Attribute
    {
    }
}";
            var comp = CreateNullableCompilation(source);
            comp.VerifyDiagnostics();
        }
```

I will confirm whether this behavior is desirable.

Tagging @jskeet @svick FYI

**Comment by jcouv** [comment]
Confirmed with Mads that the current behavior of `[NotNull]` on by-value and `in` parameters is what we intended. That means we don't need a `[DoesNotReturnWhenNull(parameterName)]` attribute anymore.
I'll go ahead and close this issue.

**Comment by jskeet** [comment]
@jcouv: Is it worth me filing an issue to ask for that to be clarified in docs? At the moment the docs are just "Specifies that an output is not null even if the corresponding type allows it." which really don't make this clear.

If the attributes are already planned to undergo a docs refresh, it's not worth it - but I'm happy to file the issue if it's useful.

**Comment by jcouv** [comment]
Let's adjust the LDM notes to clarify. Filed a PR on csharplang (linked)

**Comment by Sram-Tools** [comment]
DoesNotReturnWhenNull is highly apprecdiated in case of Requires.NotNull(arg, argName)
VS warns about possible nullability after that line. That's nasty.

**Comment by XETH-Coding** [comment]
DoesNotReturnWhenNull is highly apprecdiated in case of Requires.NotNull(arg, argName)
VS warns about possible nullability after that line. That's nasty.

**Comment by RikkiGibson** [comment]
Could you please clarify what assembly and namespace the `Requires.NotNull` method comes from and what target framework you are using?

**Comment by CleanCodeX** [comment]
> Could you please clarify what assembly and namespace the `Requires.NotNull` method comes from and what target framework you are using?

I am using core 3.1 and 5.0.
Requires.NotNull is written by myself. To my knowledge there is nothing similar public in the .net framework.
It's a method which does the following:

`public static void NotNull(object? arg, string argName)
{
     if (arg is null)
          throw new ArgumentNullException(argName, Strings.ArgumentMustNotBeNull);
}`

**Comment by RikkiGibson** [comment]
In this case you should just add the `[NotNull]` attribute to the `arg` parameter. This denotes that the method only ever returns normally if the argument passed for that parameter is non-null, and the compiler recognizes this by giving the argument a not-null state after a call to the method.

**Comment by CleanCodeX** [comment]
> In this case you should just add the `[NotNull]` attribute to the `arg` parameter. This denotes that the method only ever returns normally if the argument passed for that parameter is non-null, and the compiler recognizes this by giving the argument a not-null state after a call to the method.

That makes no sense, because the argument can indeed be null. The compiler should _not_ warn if a possible null value might has been passed as argument. 
Requires.NotNull should prevent execution if the argument is null and there is no proper way (yet) to do that despite Debug.Asert(variable != null).

A attribute DoesNotReturnWhenNull(nameof(param)) would fit all needs.

**Comment by RikkiGibson** [comment]
The NotNullAttribute is a postcondition, as opposed to the AllowNullAttribute which is a precondition specifying that null may be provided as an argument. Please see https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis.

Your feedback on the design is appreciated, but since it has shipped, it is unlikely to change.

**Comment by CleanCodeX** [comment]
Well thanks, didn't know that NotNull is a postcondition. The wording of that attribute confuses instead of following the principle of least surprise (at least to me).
Something like DoesNotReturnWhenNull(name) leaves no space for interpretation.
(maybe without param when placed on param directly instead on method)

"Your feedback on the design is appreciated, but since it has shipped, it is unlikely to change."

Actually this would not need a change rather than an addition in adding one further attribute. There wouldn't be a breaking change at all or a any chage in existing behavior.
Is there any reason why NotNull and DoesNotReturnWhenNull cannot coexist?

Anyway, would you place it on the agenda for a later release?

**Comment by jcouv** [comment]
> The wording of that attribute confuses instead of following the principle of least surprise (at least to me).

We can adjust the wording in documentation to reduce surprise, but there is no escaping that to understand nullability attributes one must understand which ones are pre-conditions and which ones are post-conditions. I'd suggest filing a docs bug on NotNull if that's not clear enough (or even better a docs PR). 

Adding a new attribute for a scenario that's already addressed in the feature design will not be worthwhile in terms of value vs. complexity. Thanks

**Comment by CleanCodeX** [comment]
I think clarity should always outweigh some sort of slightly increased "complexity". Updating the NotNull docu (in Code) should only be the minimum effort to provide a quick fix for the likely missunderstanding.
Despite the purpose/usefulness of NotNull placed on every not nullable param it looks much cleaner (and imho nicer) if you can place the attribute on the method itself (as the similar others do)

Compare:
[DoesNotReturnWhenNull("paramA", "paramB", "paramC")]
void MethodA(string? paramA, bool? paramB, int? paramC)() 
{}

Looks imho way better than streching the param list:
void MethodA([NotNull] string? paramA, NotNull] bool? paramB, NotNull] int? paramC)() 
{}

Again: NotNull attribute indicates in no way that the method does not return if one of its params is null. No chance. Whereas DoesNotReturnWhenNull makes it clear.

**Comment by scottdorman** [comment]
There is already a `[DoesNotReturnIf(bool)]` attribute that can be applied to a parameter or a `[DoesNotReturn]` attribute which can be applied to the method. Would one of these, in combination with the `[NotNull]` attribute on the parameter, not be enough?

I do agree that the naming of `[NotNull]` does not make it clear that it's a post-condition.

**Comment by CleanCodeX** [comment]
> There is already a `[DoesNotReturnIf(bool)]` attribute that can be applied to a parameter or a `[DoesNotReturn]` attribute which can be applied to the method. Would one of these, in combination with the `[NotNull]` attribute on the parameter, not be enough?
> 
> I do agree that the naming of `[NotNull]` does not make it clear that it's a post-condition.

[DoesNotReturnIf(bool)] only applies for bool params not nullable params.
the attribute [DoesNotReturn] suggest that the method does not return at all. Disregarding whether a param is null or not.

If the claim is that the .net framework should follow the least surprise principle, then NotNull (in the meaning of DoesNotReturnWhenNull) is insufficient. 

Despite that it seems obvious to me, that adding an attribute with a clearer name (which can coexist) is a better solution (I don't say it's the best) than having a NotNull attribute with unclear, overloaded meanings (which not only beginners don't know).
